### DISCUSSION TITLE
How to create generic type for Control, UseFormSetValue that work.

### DISCUSSION THREAD

**Discussion by Junaid300** [react-hook-form/react-hook-form#10873] [General]
I have a common field that is going to use in multiple form . I want to avoid duplication and will to use that common field in my multiple component. I want to create generic that accept control and setValue of different kind of form.
For example i have `country listing Select field`. Two forms that is for `profile` and `Post` , in these two form i gonna use `Country select field` ,
Now i have two different forms and in these form i have different control and setValues. How can i give type to props of `Country Select Field` that will able to accept  props from both forms and will able to set of get value.
Mainly i want to know how can i give type for a common field that gonna use in different kind of forms.

**Comment by Moshyfawn** [comment]
Take a look at #8618 to see if it's what you're after

**Reply by Junaid300** [reply]
I think i am not able to example my problem properly. Two forms, both forms has totally different types and useFormHook. I have a common field and i want i can use my common field in both forms without getting typescript errors for control or setValue. 
I got this from chat gpt but still confuse. 
```
import React from 'react';
import { Control, FieldValues } from 'react-hook-form';

type CommonFieldProps<T extends FieldValues> = {
  control: Control<T>;
  name: string;
  label?: string;
  // Add any other common props you want to support
};

function CommonInputField<T extends FieldValues>({ control, name, label, ...rest }: CommonFieldProps<T>) {
  return (
    <div>
      {label && <label>{label}</label>}
      <input
        {...rest}
        name={name}
        ref={control.register}
      />
    </div>
  );
}

type Form1Data = {
  // Define the fields specific to Form 1
  field1: string;
  field2: number;
};

type Form2Data = {
  // Define the fields specific to Form 2
  fieldA: string;
  fieldB: boolean;
};

function Form1({ control }: { control: Control<Form1Data> }) {
  return (
    <form>
      <CommonInputField control={control} name="field1" label="Field 1" />
      <CommonInputField control={control} name="field2" label="Field 2" />
      <button type="submit">Submit Form 1</button>
    </form>
  );
}

function Form2({ control }: { control: Control<Form2Data> }) {
  return (
    <form>
      <CommonInputField control={control} name="fieldA" label="Field A" />
      <CommonInputField control={control} name="fieldB" label="Field B" />
      <button type="submit">Submit Form 2</button>
    </form>
  );
}

function App() {
  return (
    <div>
      <h1>Form 1</h1>
      <Form1 control={/* pass control for Form 1 */} />

      <h1>Form 2</h1>
      <Form2 control={/* pass control for Form 2 */} />
    </div>
  );
}

export default App;
```

If i implement this, i get error for names . `Type 'string' is not assignable to type 'Path<T>'`

**Reply by akhalinem** [reply]
Hey @Junaid300 I'm facing the same issue. Here I made a demo to represent the issue: 

<img width="1088" alt="Screenshot 2023-11-25 at 1 21 12 AM" src="https://github.com/react-hook-form/react-hook-form/assets/33521936/c514ee7c-96f0-40a7-ba7a-4feb72d3cf0b">

Link to the demo: https://codesandbox.io/p/devbox/rhf-generic-setvalue-qy6kml

**Reply by akhalinem** [reply]
I wonder if there is an elegant type-safe solution to this issue.

**Reply by akhalinem** [reply]
@Moshyfawn Could you please provide some thoughts on this as well?

**Reply by Junaid300** [reply]
Sorry brother, i am not sure about it.

**Reply by ocruze** [reply]
Have you guys found any solution? I have the same problem :(

**Reply by Junaid300** [reply]
This is how i was able to solve this problem
Here is my Generic Type
```
type NetworkCountriesFieldProps<T extends FieldValues> = {
  control: Control<T>;
  setValue: UseFormSetValue<T>;
  watchKey: Path<T>;
  stateWatchKey?: Path<T>;
  documentWatchKey?: Path<T>
};
```
Here is my component that is using generic type
```export function NetworkCountriesField<T extends FieldValues>({
  control,
  setValue,
  watchKey,
  stateWatchKey,
  documentWatchKey,
  
}: NetworkCountriesFieldProps<T>) {
return(
 <CountryStatesSidebar
    countries={CountriesList}
    selectedCountries={countries}
    selectedStates={states}
    onConfirm={(countries: SelectedCountry[], states?: SelectedState[]) => {
    setValue(watchKey, countries as PathValue<T, Path<T>>, { shouldDirty: true });
    setValue(stateWatchKey as Path<T>, states as PathValue<T, Path<T>>, { shouldDirty: true });
    setCountryToggle(false);
    }}
     onCancel={() => setCountryToggle(false)}
 />
)
}
```
This is how i used the generic type

```
 <NetworkCountriesField<AddNewNetworkFormType>
    control={control}
    setValue={setValue} />
```

**Reply by ocruze** [reply]
Thank you for your reply. But I'm actually looking for a way to avoid `as PathValue<T, Path<T>>` everywhere. If anyone finds a way to do this, please let me know.

**Reply by Junaid300** [reply]
I was unable to find any alternate solution. And i completed my implementation with this solution

**Comment by aspirisen** [comment]
Hi, I created a PR that demonstrates a Lens approach for building reusable components with TypeScript support https://github.com/react-hook-form/react-hook-form/pull/12284
